chain/chain/src/store.rs

```diff
@@ -1818,8 +1818,16 @@ impl<'a> ChainStoreUpdate<'a> {
         self.chain_store_cache_update.outcome_ids.insert((*block_hash, shard_id), outcome_ids);
     }
 
-    pub fn save_trie_changes(&mut self, trie_changes: WrappedTrieChanges) {
+    // fix me: catch error everywhere
+    pub fn save_trie_changes(&mut self, trie_changes: WrappedTrieChanges) -> Result<(), Error> {
+        // Pop deleted trie nodes from the shard cache.
+        // TODO: get rid of creating `StoreUpdate`
+        let mut store_update = self.store().store_update();
+        trie_changes.deletions_into(&mut store_update);
+        store_update.update_cache()?;
```

Comment:
:thinking: this feels wrong to me. As far as I understand, `save_` family of methods isn't supposed to actually modify any state. Rather, it saves changes themselves to be applied later during commit. It seems that what we want to do here is: 

```diff
diff --git a/chain/chain/src/store.rs b/chain/chain/src/store.rs
index 9ea30e8a0..8d9a3c83a 100644
--- a/chain/chain/src/store.rs
+++ b/chain/chain/src/store.rs
@@ -1820,12 +1820,6 @@ impl<'a> ChainStoreUpdate<'a> {
 
     // fix me: catch error everywhere
     pub fn save_trie_changes(&mut self, trie_changes: WrappedTrieChanges) -> Result<(), Error> {
-        // Pop deleted trie nodes from the shard cache.
-        // TODO: get rid of creating `StoreUpdate`
-        let mut store_update = self.store().store_update();
-        trie_changes.deletions_into(&mut store_update);
-        store_update.update_cache()?;
-
         self.trie_changes.push(trie_changes);
         Ok(())
     }
@@ -2786,6 +2780,7 @@ impl<'a> ChainStoreUpdate<'a> {
         for mut wrapped_trie_changes in self.trie_changes.drain(..) {
             wrapped_trie_changes.insertions_into(&mut store_update);
+            wrapped_trie_changes.deletions_into(&mut store_update);
             wrapped_trie_changes.state_changes_into(&mut store_update);
 
             if self.chain_store.save_trie_changes {
                 wrapped_trie_changes
```

Reply:
I can't remember why I did this... perhaps a legacy from some other idea

Reply:
Well, tests helped us to remember.
We want to apply deletions to cache, but *we don't want to apply them to store*. Straightforward `deletions_into` call means that refcount of node is decremented and it will be likely removed from store right now. Instead, we want to keep node in the store until GC happens.